### DISCUSSION TITLE
Edge-based WebView2 - Details please

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#1170] [Q&A] [answered]
@ajwfrost : Can you please give some details on the following new feature of new AIR SDK Release (AIR 33.1.1.620)?

`AIR-4985:
Implementation of Edge-based WebView2 for Windows apps with ADT support for application.windows.UseWebView2`

**Comment by D5CN** [comment]
is great to get news about WebView update. need details too.

**Comment by ajwfrost** [comment, marked as answer]
Two things are needed if you want to use the Edge WebView2 component for StageWebView:

1) the application descriptor file update:
```
<windows>
  <UseWebView2>true</UseWebView2>
</windows>
```

2) the WebView2 Runtime needs to exist on the target computer:
https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section


That's basically it...  we're still going through some of the test cases, and have spotted a couple of minor things like the order of events (navigation complete vs page complete) that are different between the two variants. Plus we've not yet added the draw-to-bitmapdata functionality - not sure if anyone actually uses that..?!

Regarding the WebView2 Runtime dependency, I'd be interested in your thoughts on how that should be handled. There are a few options from Microsoft, described in detail at:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution
My concern are around the license restrictions and requirements that are shown when you try to get the link for the Evergreen Bootstrapper. So I would actually prefer just to direct folk to the website and download section as above, and ask them to install it themselves... but if you were happy with the agreement and could then distribute a link in your app, you could add some manual component to download and run it. 

Let us know how it goes, I wouldn't suggest deploying this to customers quite yet as there may be some teething issues with it..

thanks

**Reply by D5CN** [reply]
> Two things are needed if you want to use the Edge WebView2 component for StageWebView:
> 
> 1. the application descriptor file update:
> 
> ```
> <windows>
>   <UseWebView2>true</UseWebView2>
> </windows>
> ```
> 
> 1. the WebView2 Runtime needs to exist on the target computer:
>    https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section
> 
> That's basically it... we're still going through some of the test cases, and have spotted a couple of minor things like the order of events (navigation complete vs page complete) that are different between the two variants. Plus we've not yet added the draw-to-bitmapdata functionality - not sure if anyone actually uses that..?!
> 
> Regarding the WebView2 Runtime dependency, I'd be interested in your thoughts on how that should be handled. There are a few options from Microsoft, described in detail at:
> https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution
> My concern are around the license restrictions and requirements that are shown when you try to get the link for the Evergreen Bootstrapper. So I would actually prefer just to direct folk to the website and download section as above, and ask them to install it themselves... but if you were happy with the agreement and could then distribute a link in your app, you could add some manual component to download and run it.
> 
> Let us know how it goes, I wouldn't suggest deploying this to customers quite yet as there may be some teething issues with it..
> 
> thanks

Thanks for your answer. I installed WebView2 SDK, and updated the application descriptor file.
but crush again, the source code is:

var webview:StageWebView = new StageWebView(true);
webview.stage = this.stage;
webview.viewPort = new Rectangle(0, 0, 500,500);
webview.loadURL("http://www.d5power.com/");

what I am lost..

**Reply by 2jfw** [reply]
> That's basically it... we're still going through some of the test cases, and have spotted a couple of minor things like the order of events (navigation complete vs page complete) that are different between the two variants. Plus we've not yet added the draw-to-bitmapdata functionality - not sure if anyone actually uses that..?!
> 
> Regarding the WebView2 Runtime dependency, I'd be interested in your thoughts on how that should be handled. There are a few options from Microsoft, described in detail at:
> https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution
> My concern are around the license restrictions and requirements that are shown when you try to get the link for the Evergreen Bootstrapper. So I would actually prefer just to direct folk to the website and download section as above, and ask them to install it themselves... but if you were happy with the agreement and could then distribute a link in your app, you could add some manual component to download and run it.
> 
> Let us know how it goes, I wouldn't suggest deploying this to customers quite yet as there may be some teething issues with it..


DrawToBitmap could be usefull for a small live preview/thumbnail-like display of the website, so basically what you are about to open.
Other than that, I could not imagine where else I would use that.

Regarding the version, we would use the fixed runtime distribution mode and put it in the installer.
Customer should have to do as little as possible to get things run or working.

**Reply by itlancer** [reply]
@ajwfrost 
> Plus we've not yet added the draw-to-bitmapdata functionality - not sure if anyone actually uses that..?!

Yes, this functionality needed)

> Regarding the WebView2 Runtime dependency, I'd be interested in your thoughts on how that should be handled. There are a few options from Microsoft, described in detail at:
> https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution
> My concern are around the license restrictions and requirements that are shown when you try to get the link for the Evergreen Bootstrapper. So I would actually prefer just to direct folk to the website and download section as above, and ask them to install it themselves... but if you were happy with the agreement and could then distribute a link in your app, you could add some manual component to download and run it.

It ok but I think we need a way at runtime to check does WebView2 installed (supported) and can be used by AIR.

**Reply by ajwfrost** [reply]
Just wondering about this, as we don't want to have a Windows-specific ActionScript API that knows about IE/Edge.. is it safe to assume that if you put the app descriptor setting to UseWebView2, we should only try to create a WebView2, and fail if it's not present? Or is there any benefit in providing a fallback mode?

If we're just going to create whichever option was requested per the descriptor file, then we can update that isSupported to check based on the availability of the DLLs and registry entries about it.

If we want to provide a programmatic option to choose one or the other, then it becomes a bit trickier and we'd need a new API :-(

(and I'll ask someone to start work on that draw-to-bitmapdata piece....)

thanks

**Reply by 2jfw** [reply]
@ajwfrost I totally missed the **Release Notes PDF** containing information, this time. Maybe additionally add a link to the PDF in the "Release Notes" section / overview directly, rather than only having it on the "Download" part as a small side note?

**Reply by 2jfw** [reply]
@ajwfrost Regarding Fallback, from our side, there is no necessity to have a fallback, because the fallback (/old webview) is way too outdated to consider it

**Reply by itlancer** [reply]
@ajwfrost 
>If we're just going to create whichever option was requested per the descriptor file, then we can update that isSupported to check based on the availability of the DLLs and registry entries about it.

Updated isSupported method depends on descriptor setting will be good enough.
"New API" for webframe functionality should be considered as a next step that could solve current problems and restrictions of `HTMLLoader` and `StageWebView`. Some of them:
https://github.com/airsdk/Adobe-Runtime-Support/issues/1106#issuecomment-911616885
https://github.com/airsdk/Adobe-Runtime-Support/issues/143
https://github.com/airsdk/Adobe-Runtime-Support/issues/337
https://github.com/airsdk/Adobe-Runtime-Support/issues/267
https://github.com/airsdk/Adobe-Runtime-Support/issues/223
https://github.com/airsdk/Adobe-Runtime-Support/issues/214

**Comment by D5CN** [comment]
and in the console I got message:

Exception in debugger: flash.tools.debugger.NotConnectedException: 
	at flash.tools.debugger.concrete.PlayerSession.isSuspended(PlayerSession.java:311)
	at flash.tools.debugger.threadsafe.ThreadSafeSession.isSuspended(ThreadSafeSession.java:174)
	at com.as3mxml.vscode.SWFDebugSession$SessionRunner.run(SWFDebugSession.java:252)
	at java.lang.Thread.run(Unknown Source)

**Reply by 2jfw** [reply]
Just making sure - did you update to latest AIR Version?

**Reply by D5CN** [reply]
> Just making sure - did you update to latest AIR Version?

Yes, that is the first step... ^^

**Reply by ajwfrost** [reply]
Can you put a try/catch around that block of code to see whether any error is thrown?
And can you please check whether you have a file `WebView2Loader.dll` on your machine, and if so where it is? (I have a few by the looks of things but think it will be picking up from `C:\Windows\CCM\ClientUX`)
Plus - are you running in the default 32-bit mode, or have you got it launching in 64-bit?

**Reply by D5CN** [reply]
Thanks! i'll check it.

**Reply by ilmarfreitas** [reply]
Hello.
I am also not having success using StageWebView with WebView2. I couldn't find the WebView2Loader.dll file on my machine, which is a Windows 10 64bits where I have Edge and Microsoft WebView2 installed (both in version 93.0.961.47).
Thanks.

**Reply by ajwfrost** [reply]
Okay thanks @ilmarfreitas  - so on digging into this deeper, it looks like we do have the option of statically linking in the function from Microsoft that's used for finding/loading the webview2 component. We'll try that and perhaps also add in some debugging output in the short term in case it helps find the issues.

**Reply by D5CN** [reply]
Hi, I tried again.
There is no WebView2Loader.dll file on my machine too. My OS is Windows 10 64bits, have Edge and WebView2 SDK installed.Edge version is 93.0.961.47,WebView2 version is 93.0.961.52. I tried with description 
`<architecture>64</architecture>`
but the result was same. It crushed and try/catch can not work.

But when I use
```
<Windows> // it is <windows> in official document
		<UseWebView2>true</UseWebView2>
</Windows>

```
It stops crushed. but webview is not worked. i got error message:
2004 ArgumentError: Error #2004: One of the parameters is invalid.
	at flash.media::StageWebView/set stage()
	at Main/loadCore()[j:\Client\AS3\Project\SVN_109\AS3\D5PowerBitmapUI\D5BitmapUIEditor\src_main\Main.as:285] Error #2004: One of the parameters is invalid.

if I skeep this code,error was:
2004 ArgumentError: Error #2004: One of the parameters is invalid.
	at flash.media::StageWebView/set viewPort()
	at Main/loadCore()[j:\Client\AS3\Project\SVN_109\AS3\D5PowerBitmapUI\D5BitmapUIEditor\src_main\Main.as:286] Error #2004: One of the parameters is invalid.

I'll keep trying,and post new message here.Thanks

**Reply by ajwfrost** [reply]
If you use `<Windows>` then we probably won't work out that there is a request for the newer webview. So I would have thought the behaviour there would be the same as if the whole section was missing. At least, I believe that this is all case-sensitive...

We'll put together a workaround for the lack of WebView2Loader.dll and post it up here with some Scout-based debugging.

thanks

**Reply by ajwfrost** [reply]
May be a slight delay on this, as the WebView2Loader is meant to be present from what we tell - but there is a static version of this that we should be able to link against instead. However there is a problem with using that static lib currently, an issue is already open with the WebView team about that. We'll see if there are any other options/workarounds that we can do...

thanks

**Reply by ilmarfreitas** [reply]
Hi @ajwfrost.Can you tell me which version of WebView2Loader.dll we should use? I will install Visual Studio to see the Visual Studio Installer will place the WebView2Loader.dll on my machine. Should it be located in the \windows or \windows\system32 folder?

**Reply by ilmarfreitas** [reply]
Hello @ajwfrost.

I installed Visual Studio 2019 and found WebView2Loader.dll version 1.0.864.35. I copied WebView2Loader.dll (32bits version) to c:\windows folder and my test application started working using WebView2. It did not work when trying to use the 64-bit WebView2Loader.dll.

Now that we know how to proceed with WebView2Loader.dll, we can place its installation inside our installation package.

Thanks

![htm5test](https://user-images.githubusercontent.com/56614220/134027717-99be2c2e-de18-4cd5-adb9-967838be6f68.png)
.

**Reply by ajwfrost** [reply]
Hi @ilmarfreitas  - that sounds good, but yes there's a potential issue because we're calling LoadLibrary in order to do this, and this won't work if the library doesn't have the same architecture (x86/x64) as the AIR runtime.

The "proper" way of doing things it seems is to use the static loader library so it doesn't use the .dll file - but this is causing problems with the linker. We may be able to do the same sort of things manually ourselves though: all this library/utility seems to do is to locate the webview itself and load in this library and then create the object. We may be able to do the same by checking the registry for the location of the installed webview, and then there are two instances of these e.g.
`C:\Program Files (x86)\Microsoft\EdgeWebView\Application\93.0.961.52\EBWebView\x86\EmbeddedBrowserWebView.dll`

So we'll see if we can get that working as a better way to find/load the actual webview component....


thanks

**Reply by ilmarfreitas** [reply]
Hi ajwfrost.

If we really don't have the dependency on WebView2Loader.dll it will be a better solution. Let's wait for the release of the new solution then.

Thanks.

**Reply by ajwfrost** [reply]
@ilmarfreitas @D5-Howard  are you able to try using the DLL from the attached zip and see whether you can now create the StageWebView component? There should be some extra debug come out via Scout if that helps..
thanks
[webview2_test.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7205918/webview2_test.zip)

**Reply by D5CN** [reply]
OK,i'll try it.

**Reply by D5CN** [reply]
> @ilmarfreitas @D5-Howard are you able to try using the DLL from the attached zip and see whether you can now create the StageWebView component? There should be some extra debug come out via Scout if that helps..
> thanks
> [webview2_test.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7205918/webview2_test.zip)

Yes! It work!!
I copy dll files to AIRSDK/runtimes/air/ to overwrite the old version.
then StageWebView is work well. both in 32 bit and 64 bit.

maybe WebView2Loader.dll is not necessary, i delete it from Windows and Windows/system32, but StageWebView is still work well.

**Reply by ilmarfreitas** [reply]
Hello @ajwfrost. Here it also worked perfectly without using WebView2Loader.dll, just like @D5-Howard said. Excellent improvement you have made. Congratulations!

Thanks

**Reply by ajwfrost** [reply]
Great, thanks both for the confirmations, we can go with this approach for now then and hopefully move to use the 'official' statically linked webviewloader library once MSFT have sorted that out (or if/when we switch to a different Visual Studio version...)

**Reply by itlancer** [reply]
@ajwfrost
"Old" `StageWebView` implementation doesn't work with AIR 33.1.1.620 for Windows devices even with your patch https://github.com/airsdk/Adobe-Runtime-Support/discussions/1170#discussioncomment-1364163
`var stageWebView:StageWebView = new StageWebView(true);`
cause error:
```
Error: Error #1001: The method StageWebView is not implemented.
	at flash.media::StageWebView/init()
	at flash.media::StageWebView()
```
with 
```
<windows>
	  <UseWebView2>false</UseWebView2>
</windows>
```
or without this section at all.

Without your patch we have error mentioned here: https://github.com/airsdk/Adobe-Runtime-Support/discussions/1170#discussioncomment-1350359

**Reply by ajwfrost** [reply]
Thanks @itlancer  - looks like a conflicting definition meant that only one of the webview implementations would work at once, depending on the order of the include files! This will be sorted out for the proper release...

**Comment by D5CN** [comment]
![dbebd7ad0d053d78b540387fde53276](https://user-images.githubusercontent.com/16438099/134271424-c8924ff6-969b-47f5-9afa-9e631eeca923.png)

**Comment by ajwfrost** [comment]
Just FYI regarding the `drawViewPortToBitmapData` method -> this will need to be asynchronous so when the function returns, it will not have drawn anything.. but we can then update the bitmap data object once we've got the preview available from the WebView2 component.

Thinking we should probably add an extra event to the StageWebView object so that you will know when the provided BitmapData is valid. You can always cancel the operation by calling `dispose` on the BitmapData object...

**Reply by itlancer** [reply]
@ajwfrost 
Or may be some new method `drawViewPortToBitmapDataAsyncRequest` with new event that contains `BitmapData`.

**Reply by ajwfrost** [reply]
... if we're going to change the API anyway, would it be okay to instead have `drawViewPortToLoaderAsync` or just `drawViewPortToByteArrayAsync` with a requested format (PNG or JPG) that can then be passed to Loader.loadBytes()?  Since the implementation of WebView allows us to request one of these formats, and then just gives us the bytes.

If you'd prefer BitmapData, it's not a big deal as we can then do the PNG decode into this, but if folk were saving snapshots to disk as image files, it may make it simpler if we just dispatch the raw image data....

**Reply by itlancer** [reply]
@ajwfrost 
For me it better to have new method to get `BitmapData` cause we use it for render/composing, not just saving as a file. PNG/JPG data we can create by ourselves via `BitmapData::encode()` method.

**Reply by ajwfrost** [reply]
Fair enough!

To notify the the bitmap data is ready ... the `complete` event is already being used by the `StageWebView` object so I think we need to send a new one - I think you're proposing a new event subclass for this so that we can dispatch the bitmap data within this? e.g. `WebViewDrawEvent`. Then we could have `drawViewPortToBitmapDataAsync` (and you can pass in a `BitmapData` object to avoid us having to create a new one if you want..) and it would require you to have set up a listener in order to call it... the event is then dispatched and you can get at the `BitmapData` as a member of this.

In case it's helpful to others, we could mirror this with `drawViewPortToByteArrayAsync` where you can request a PNG or JPEG file in the same way. Avoids needing us to decode the PNG that Windows gives us and then re-encode it from a later ActionScript request ..

**Reply by itlancer** [reply]
> To notify the the bitmap data is ready ... the complete event is already being used by the StageWebView object so I think we need to send a new one - I think you're proposing a new event subclass for this so that we can dispatch the bitmap data within this? e.g. WebViewDrawEvent. Then we could have drawViewPortToBitmapDataAsync (and you can pass in a BitmapData object to avoid us having to create a new one if you want..) and it would require you to have set up a listener in order to call it... the event is then dispatched and you can get at the BitmapData as a member of this.

Exactly!

**Reply by radubirsan** [reply]
When can we get a tutorial on how to use this?
I tried to download the latest SDK and installing the WebView from the link:
here is  my code:

import flash.events.*;
import flash.system.*;
var webView:StageWebView = new StageWebView();
webView.viewPort = new Rectangle(0, 0, stage.stageWidth, stage.stageHeight);
webView.stage = stage;
webView.addEventListener(Event.COMPLETE, complete); //event is dispatched, but with js errors
webView.loadURL("https://www.flash.pm/");

function complete(e){
	trace(StageWebView.isSupported, 999)
	}

It dose not render properly...

**Reply by ajwfrost** [reply]
The main thing you need to do is to open your application descriptor file and add the below just before the closing `</application>` line:
```
	<windows>
		<UseWebView2>true</UseWebView2>
	</windows>
```

But for your code, the other issue is the constructor:
```
var webView:StageWebView = new StageWebView();
```
The StageWebView default constructor would use the built-in WebKit engine that ships with AIR: instead of this, you want to use the 'native' i.e. platform-provided version. So if you change that to:
```
var webView:StageWebView = new StageWebView(true);
```
it should work.


thanks

**Reply by radubirsan** [reply]
Thank you, it works! ( make sure you use AIR SDK 33.1.1.633 ) 
Looking forward to a bundling solution example ;)

**Reply by herrfinke** [reply]
@ajwfrost is there an update with the `drawViewPortToBitmapDataAsync()` method that I missed? We could really use that method, because the `drawViewPortToBitmap` is increadibly slow when having websites which include animated content in FullHD or even larger resolutions.
Or would that method not be required anymore if Air 51 supports the webview to be part of the normal UI-View (and could be overlayed by UI-Components)?

**Reply by herrfinke** [reply]
Hi, I found the method `drawViewPortAsync` is working, and will lead to the dispatch of a `WebViewDrawEvent.WEBVIEW_DRAW_COMPLETE`
Nonetheless: it would be way cooler to have the interactive (and realtime) webview in the UI-Tree. Looking forward to that release :-)

**Reply by ajwfrost** [reply]
Hi - the 'async' method here is actually required when you use the Edge WebView2 variant, because their API for capturing the screen this is itself asynchronous.. in the case of the older IE-based webview, this is embedded via ActiveX and afaik would need to be captured within the same thread as the normal rendering.

Having the webview within the normal display list would require us to be able to get at the pixels, and for Edge WebView2 that isn't going to be possible.. it may be possible to build external browser versions to render into a buffer that can then be used as a displaylist object, I'm not sure how performant this might be but that's the kind of thing we're trying to enable with a new ANE function to allow external rendering into a display object..

thanks

**Comment by esidegallery** [comment]
Just gave this a go and very happy with the way it's working! What are the chances of giving access to the CoreWebView2EnvironmentOptions.AdditionalBrowserArguments property? Perhaps this could be a property in the application descriptor? This would open up a plethora of extra functionality (such as remote debugging port, which I need for my app) with no API changes.

P.S. I would also like the ability to disable the context menu via right-click / long-press.

**Reply by ajwfrost** [reply]
Thanks .. we'd been looking at ways in which these could be more customised, and will perhaps add in some new API for this sort of configuration. Updating something via the app descriptor is one way of doing things but means that the setting would be the same across the whole app, whereas it may be that you would like different behaviours at different times.

So if we provided a configuration object to the `StageWebView` constructor, rather than the two boolean args, it would allow some more extensible mechanism (i.e. if the object was dynamic). For example:
```
var config : Object = {
        additionalBrowserArguments: "--disable-gpu-vsync",
        disableContextMenu: true,
        mediaPlaybackRequiresUserAction: false
};
var swv : StageWebView = new StageWebView(config);
... etc
```

**Reply by esidegallery** [reply]
Thank you Andrew! Agreed, this makes total sense.

**Comment by Jason0247** [comment]
For hierarchical issues, I would like to add a return button on the top of the webpage, which is not achievable, is it?
Another question is, does the latest version of the AIR SDK also require the computer to install Edge in order to use it? 
Thank you! @ajwfrost

**Reply by ajwfrost** [reply]
@Jason0247 currently the StageWebView is rendered completely separately (by the OS) and then presented on the top layer of the "stage", so all native AS/displaylist-based content would appear underneath it.

The only thing I can think of for this might be to create another StageWebView with the button, and if you added this after the main webpage it should then be above it? (although that behaviour may well be OS-specific so it may change depending on whether you click in the main webpage area..)

Another option - although again, not 100% sure if this would work, and certainly it would only be for desktops - would be to create a transparent NativeWindow object and put your button on this, and then manually move this so that it's always above your main window. Bit of a hack but it may provide this sort of capability...

**Reply by Jason0247** [reply]
The window version can be implemented through a new form, which I applied three years ago.
But for mobile platforms, it is impossible to achieve this through such operations.
A good solution is to add a return in HTML and calculate the area clicked by the mouse in AIR to achieve closure, but it always feels very unfriendly.

**Reply by itlancer** [reply]
@ajwfrost 
We also looking for a way to overlay `DisplayObject` (or even `Stage3D`) over `StageWebView`.
May be one idea make sense.
What if allow to create multiple `Stage` instances (like `Stage3D`) for one `NativeWindow`? And allow to "mix" them between each other.
For example (layers from top to bottom):
- `Stage` №3
- - `Stage3D` (related to `Stage` above)
- - - Some textures using Starling
- `Stage` №2
- - `DisplayObject`
- `Stage` №1
- - `StageWebView` (related to `Stage` above)
- `Stage` №0 ("main")
- - `DisplayObject`


`Stage3D` with transparency support also will be appreciated.

**Reply by ajwfrost** [reply]
@itlancer I had been wondering the exact same thing this morning. On some platforms, this may work very well, if we can layer up the native implementations and back them by the ActionScript equivalents. I don't recollect how the Android StageWebView display is handled, but handing off the layering complexity to the OS may actually be a good idea.. assuming that they support such transparent build-up of elements like this.

We can look into this a bit, and maybe do some playing around with ideas. Makes me think we should have a "labs" style preview area for AIR.. :-)

**Reply by 2jfw** [reply]
I think we had a similar topic and ideas on this a while back - just cross-referring the older Feature Request here:
 **Possibility to overlay StageWebView** https://github.com/airsdk/Adobe-Runtime-Support/issues/1428

**Reply by ajwfrost** [reply]
Oh yes! thanks for that link. FYI that idea of rendering content into the displaylist is something that is in AIR 51.0... perhaps needs a bit of refinement and optimisation, but at the basic level of pixel buffers it is working. I would still have some hesitation about using that sort of thing for large rendered content such as web pages from an external renderer, but the layering thing may help (particularly too if we could also expose the native windowing/surface handles via the existing ANE mechanism..)

**Reply by 2jfw** [reply]
Sounds good!
So, AIR 51.0... when approx.? ;-)

**Reply by ajwfrost** [reply]
> AIR 51.0... when approx.? ;-)

Am aiming to get an initial beta for this out by the end of the year. There will be a fair number of "unfinished" parts though, the main thing is trying to work out all the different AS3 API updates that are required, since those are things that we're only looking to update once a year with the major version change.. (there is a fair amount of overhead involved when updating the major - and minor - version numbering..!)

**Comment by Jason0247** [comment]
I would like to ask a question, what browser is AIR StageWebView used on the Android system?
When I try to access some HTML using an Android browser, I also encounter various incompatibility issues, but when I use the browser on my Android phone to access it, it is normal.
Does this also have any settings like Windows?
thanks @ajwfrost

**Reply by ajwfrost** [reply]
On Android, it's a specialisation of `android.webkit.WebView`. Not sure whether the issues you're seeing may be down to compatibility/configuration settings, would be interesting to know if there are specific differences that you could isolate?

There are settings but currently none that are accessible, we were looking at providing a (generic/extendable) configuration mechanism as part of the StageWebView updates see e.g. https://github.com/airsdk/Adobe-Runtime-Support/discussions/1170#discussioncomment-2957418

thanks

**Reply by Jason0247** [reply]
I'm not sure what the specific reason is. If AIR uses `android.webkit.WebView` by default on Android systems, the incomplete display of web pages may be due to compatibility issues with `android.webkit.WebView` itself. I'm not sure if this is the reason or if there are other solutions.
Just like how the Windows platform can solve the problem by adding the following description files, is there an optional setting on the Android platform?
```
<windows>
<UseWebView2>true</UseWebView2>
</windows>
```

**Reply by ajwfrost** [reply]
We don't have any such option on Android; it just uses the WebView component directly.. but if there are configuration items/settings that could be the culprit, please let us know.. Essentially we're just using the below mechanism (I just checked to make sure we're still using the recommended class..!)
https://developer.android.com/develop/ui/views/layout/webapps/webview
